packages/runtime/runtime-utils/src/serializer.ts

```diff
@@ -27,68 +30,116 @@ export class FluidSerializer implements IFluidSerializer {
+        // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
         return !!input && typeof input === "object"
-            ? this.recursivelyReplaceHandles(input, bind)
+            ? this.recursivelyReplace(input, this.decodeValue)
             : input;
     }
 
     public stringify(input: any, bind: IFluidHandle) {
         return JSON.stringify(input, (key, value) => {
-            // If the current 'value' is not a handle, return it unmodified.  Otherwise,
-            // return the result of 'serializeHandle'.
             // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
-            const handle = !!value && value.IFluidHandle;
-            // TODO - understand why handle === false in some of our tests
-            // eslint-disable-next-line max-len
-            // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions,@typescript-eslint/no-unsafe-return
-            return handle
-                ? this.serializeHandle(handle, bind)
-                : value;
+            return value && this.encodeValue(value, bind);
         });
     }
 
     // Parses the serialized data - context must match the context with which the JSON was stringified
     public parse(input: string) {
-        // eslint-disable-next-line @typescript-eslint/no-unsafe-return
-        return JSON.parse(
-            input,
-            (key, value) => {
-                if (!isSerializedHandle(value)) {
-                    // eslint-disable-next-line @typescript-eslint/no-unsafe-return
-                    return value;
-                }
-
-                // Old documents may have handles with relative path in their summaries. Convert these to absolute
-                // paths. This will ensure that future summaries will have absolute paths for these handles.
-                const absolutePath = value.url.startsWith("/")
-                    ? value.url
-                    : generateHandleContextPath(value.url, this.context);
-                return new RemoteFluidObjectHandle(absolutePath, this.root);
-            });
+        // Note: 'decodeValue()' is tolerant of undefined/null, so no need to check here.
+        return JSON.parse(input, (key, value) => this.decodeValue(value));
     }
 
-    // Invoked by `replaceHandles()` for non-null objects to recursively replace IFluidHandle references
-    // with serialized handles (cloning as-needed to avoid mutating the original `input` object.)
-    private recursivelyReplaceHandles(
+    // If the given 'value' is an IFluidHandle, returns the encoded IFluidHandle.
+    // Otherwise returns the original 'value'.
+    //
+    // Note: Caller is responsible for ensuring that 'value' is not 'undefined' or 'null'.  This
+    //       avoids redundant checking when 'encodeValue()' is applied by 'recursivelyEncode()'.
+    //
+    // (Used by 'replaceHandles()' and 'stringify()')
+    private readonly encodeValue = (value: any, bind: IFluidHandle) => {
```

Comment:
Is there any way to tell compiler that value can't be falsy here? Such that it can be more helpful in enforcing it?

Reply:
That's a good idea.  The 'Object' type means any defined/non-null value.  Although looking at it again, eliding the duplicate check is no longer important (i.e., no measurable gain).  I'll tweak the code to remove the caveat.